src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -196,8 +218,8 @@ uct_cuda_ipc_mem_add_reg(void *addr, uct_cuda_ipc_memh_t *memh,
         return UCS_ERR_NO_MEMORY;
     }
 
-    status = uct_cuda_ipc_mem_reg_push_ctx((CUdeviceptr)addr, &cuda_device,
-                                           &is_ctx_pushed);
+    status = uct_cuda_ipc_mem_reg_push_ctx((CUdeviceptr)addr, &memh->dev_num,
```

Comment:
Please update `memh` only if the method completes successfully, and keep `cuda_device` local variable.
You can update memh after `common_path` label along with adding tail to the list.